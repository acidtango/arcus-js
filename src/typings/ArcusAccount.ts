import { ArcusCurrency } from './ArcusCurrency';

export type ArcusAccount = {
  /**
   * Your company's name
   */
  name: string;
  /**
   * Your current balance
   */
  balance: number;
  /**
   * Minimum balance of you account
   */
  minimumBalance: number;
  /**
   * Your local currency
   * @pattern `/^[A-Z]{3}$/`
   */
  currency: ArcusCurrency.MXN;
};
