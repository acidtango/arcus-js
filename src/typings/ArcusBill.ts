import { ArcusBillStatus } from './ArcusBillStatus';
import { ArcusType } from './ArcusType';
import { ArcusCurrency } from './ArcusCurrency';

export type ArcusBill = {
  /**
   * Every bill type will be bill
   */
  type: ArcusType.BILL;
  /**
   * The ID of the bill
   */
  id: string;
  /**
   * The ID of the biller
   */
  billerId: number;
  /**
   * The account number of the bill */
  accountNumber: string;
  /**
   * Name on account of this bill
   */
  nameOnAccount: boolean | null;
  /**
   * The due date of the bill
   */
  dueDate: Date;
  /**
   * The balance of the bill
   */
  balance: number;
  /**
   * The balance's currency of the bill
   */
  balanceCurrency: ArcusCurrency;
  /**
   * Last time the bill was updated
   */
  balanceUpdatedAt: Date;
  /**
   * Error code if an error occurred while creating the bill. Look at API Error codes and error_message property for more information.
   */
  errorCode: string | null;
  /**
   * Error message if an error had occurred while creating the bill
   */
  errorMessage: string | null;
  /**
   * Status of the bill
   */
  status: ArcusBillStatus;
};
