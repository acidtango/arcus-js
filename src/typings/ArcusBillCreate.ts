export type ArcusBillCreate = {
  /**
   * The ID of the biller. Required
   */
  billerId: number;
  /**
   * The account number. Required
   */
  accountNumber: string;
};
