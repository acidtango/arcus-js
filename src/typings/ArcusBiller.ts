import { ArcusRange } from './ArcusRange';
import { ArcusCountryCode } from './ArcusCountryCode';
import { ArcusCurrency } from './ArcusCurrency';

export type ArcusBiller = {
  /**
   * The number of digits in the account number of the biller. This can either be a fixed number or a range. In the case of a range, the format is <min_number>..<max_number>.
   * @pattern ^\d+\.\.\d+$|^\d+$
   * @example "9..12"
   */
  accountNumberDigits: ArcusRange | number | null;
  /**
   * Boolean value specifying if the biller supports `POST /bills` or not
   */
  canCheckBalance: boolean;
  /**
   * Biller's ISO 2 digit country code
   * @example "MX"
   * @pattern `/^[A-Z]{2}$/`
   */
  country: ArcusCountryCode;
  /**
   * The 3 digit ISO currency code of the biller's country.
   * @example "MXN"
   * @pattern `/^[A-Z]{3}$/`
   */
  currency: ArcusCurrency;
  /**
   * Indicates if this bill has extended bill data.
   */
  hasXdata: boolean;
  /**
   * Specifies maximum number of hours for a successful payment to be reflected on the customerâ€™s bill. This number can vary from 0 to 72
   * @minimum 0
   * @maximum 72
   */
  hoursToFulfill: number | null;
  /**
   * Unique identifier of the biller in the Arcus system
   */
  id: number;
  /**
   * A string representing the format of the `bill_type`.
   * @example "999-999-99_99"
   */
  mask: string | null;
  /**
   * Biller's name
   */
  name: string;
  /**
   * Boolean value specifying if the name of the account owner must be passed in for payments of this type. When requires_name_on_account is true, payments that do not include the name on the account, will be rejected. See payment types for info.
   */
  requiresNameOnAccount: boolean;
  /**
   * Boolean value indicating if the biller supports paying of arbitrary amounts. When supports partial payments is false the full balance of the account must be paid, or the payment will be rejected. See payment types for info.
   */
  supportsPartialPayments: boolean;
};
