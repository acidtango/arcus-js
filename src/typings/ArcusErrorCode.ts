export enum ArcusErrorCode {
  /**
   * Invalid number of digits in account number
   */
  INVALID_NUMBER_OF_DIGITS_IN_ACCOUNT_NUMBER = 'R1',
  /**
   * Invalid Account Number
   */
  INVALID_ACCOUNT_NUMBER = 'R2',
  /**
   * Invalid Payment Amount
   */
  INVALID_PAYMENT_AMOUNT = 'R3',
  /**
   * Account number does not resolve with biller
   */
  ACCOUNT_NUMBER_DOES_NOT_RESOLVE_WITH_BILLER = 'R4',
  /**
   * Invalid Phone Number
   */
  INVALID_PHONE_NUMBER = 'R5',
  /**
   * Invalid Biller
   */
  INVALID_BILLER = 'R6',
  /**
   * Account Number is valid but cannot receive payments like this
   */
  ACCOUNT_NUMBER_IS_VALID_BUT_CANNOT_RECEIVE_PAYMENTS_LIKE_THIS = 'R7',
  /**
   * Balance has already been paid
   */
  BALANCE_HAS_ALREADY_BEEN_PAID = 'R8',
  /**
   * Unexpected error (everything entered was correct)
   */
  UNEXPECTED_ERROR = 'R9',
  /**
   * Incorrect Biller for the Provided Account Number
   */
  INCORRECT_BILLER_FOR_THE_PROVIDED_ACCOUNT_NUMBER = 'R10',
  /**
   * Invalid Payment Amount: You must pay the full amount balance
   */
  INVALID_PAYMENT_AMOUNT_UNKNOWN = 'R11', // TODO: check what this error means
  /**
   * Payment Already Made
   */
  PAYMENT_ALREADY_MADE = 'R12',
  /**
   * Name on account is required
   */
  NAME_ON_ACCOUNT_IS_REQUIRED = 'R13',
  /**
   * Payment amount too large, max amount is 1500 USD
   */
  PAYMENT_AMOUNT_TOO_LARGE_MAX_1500_USD = 'R14',
  /**
   * Account Balance is 0
   */
  ACCOUNT_BALANCE_IS_0 = 'R15',
  /**
   * Failed to make the consult, please try again later
   */
  FAILED_TO_MAKE_THE_CONSULT = 'R16',
  /**
   * Invalid Top-up amount
   */
  INVALID_TOP_UP_AMOUNT = 'R17',
  /**
   * Top-up daily limit exceeded
   */
  TOP_UP_DAILY_LIMIT_EXCEEDED = 'R18',
  /**
   * Payments with this biller are currently unavailable
   */
  PAYMENTS_WITH_THIS_BILLER_ARE_CURRENTLY_UNAVAILABLE = 'R19',
  /**
   * Lookup time out with biller
   */
  LOOKUP_TIME_OUT_WITH_BILLER = 'R20',
  /**
   * Configuration issue between us and our biller
   */
  CONFIGURATION_ISSUE_BETWEEN_US_AND_OUR_BILLER = 'R21',
  /**
   * Biller maintenance in progress, please try again later
   */
  BILLER_MAINTENANCE_IN_PROGRESS = 'R22',
  /**
   * No transaction to reverse
   */
  NO_TRANSACTION_TO_REVERSE = 'R23',
  /**
   * Timeout from biller
   */
  TIMEOUT_FROM_BILLER = 'R24',
  /**
   * Error with performing the cancellation
   */
  ERROR_WITH_PERFORMING_THE_CANCELLATION = 'R25',
  /**
   * Reversal not supported for this biller
   */
  REVERSAL_NOT_SUPPORTED_FOR_THIS_BILLER = 'R26',
  /**
   * Overdue Bill
   */
  OVERDUE_BILL = 'R27',
  /**
   * Invalid Account Number: Sky account numbers should start with 401 or 501 or 601
   */
  INVALID_ACCOUNT_NUMBER_SKY = 'R28',
  /**
   * The biller returned an error. Please check the account number.
   */
  THE_BILLER_RETURNED_AN_ERROR = 'R29',
  /**
   * Different company
   */
  DIFFERENT_COMPANY = 'R30',
  /**
   * Phone number is no longer active
   */
  PHONE_NUMBER_IS_NO_LONGER_ACTIVE = 'R31',
  /**
   * Invalid Payment Amount: Sky minimum payment is $ 185 MXN
   */
  INVALID_PAYMENT_AMOUNT_SKY = 'R32',
  /**
   * Fraud suspected
   */
  FRAUD_SUSPECTED = 'R33',
  /**
   * Invalid Currency
   */
  INVALID_CURRENCY = 'R34',
  /**
   * Temporary communication error with carrier. Please try again
   */
  TEMPORARY_COMMUNICATION_ERROR_WITH_CARRIER = 'R35',
  /**
   * Blocked attempt to make a possible duplicate payment. Requests for this account number with this payment amount are blocked for 1 hour
   */
  BLOCKED_ATTEMPT_TO_MAKE_A_POSSIBLE_DUPLICATE_PAYMENT = 'R36',
  /**
   * Recurring payments enabled for this account
   */
  RECURRING_PAYMENTS_ENABLED_FOR_THIS_ACCOUNT = 'R37',
  /**
   * Invalid Payment Amount: You must pay at least the actual account balance
   */
  INVALID_PAYMENT_AMOUNT_UNKNOWN_2 = 'R41', // TODO: check what this error means
  /**
   * Account Number is valid, but is postpaid and doesn’t have balance yet
   */
  ACCOUNT_NUMBER_IS_VALID = 'R42',
  /**
   * Invalid Barcode Format. Please the payment with the account number
   */
  INVALID_BARCODE_FORMAT = 'R43',
  /**
   * Invalid Customer Status. This customer is required to show in person at the Biller location
   */
  INVALID_CUSTOMER_STATUS = 'R44',
  /**
   * The maximum number of payments for this account on this day was reached
   */
  THE_MAXIMUM_NUMBER_OF_PAYMENTS_FOR_THIS_ACCOUNT_ON_THIS_DAY_WAS_REACHED = 'R45',
  /**
   * Can't get balance. This customer is required to show in person at the Biller location.
   */
  CANT_GET_BALANCE = 'R46',
  /**
   * No connection with the billerUnable to perform operation, biller returned an unexpected error
   */
  NO_CONNECTION_WITH_THE_BILLERUNABLE_TO_PERFORM_OPERATION = 'R47',
  /**
   * Unable to perform operation, biller returned an unexpected error
   */
  UNABLE_TO_PERFORM_OPERATION = 'R48',
  /**
   * Catchall error
   */
  CATCHALL_ERROR = 'R99',
  /**
   * Insufficient balance. Make a call to /account to get your current balance
   */
  INSUFFICIENT_BALANCE = 'R100',
  /**
   * Biller doesn’t support balance lookup
   */
  BILLER_DOESNT_SUPPORT_BALANCE_LOOKUP = 'R101',
  /**
   * Invalid or missing parameters
   */
  INVALID_OR_MISSING_PARAMETERS = 'R102',
  /**
   * Transaction already reversed
   */
  TRANSACTION_ALREADY_REVERSED = 'R103',
  /**
   * Invalid Subscription ID. Please verify the ID that you passed or create a new subscription
   */
  INVALID_SUBSCRIPTION_ID = 'R104',
  /**
   * Only same day reversals are allowed
   */
  ONLY_SAME_DAY_REVERSALS_ARE_ALLOWED = 'R105',
  /**
   * Payment amount too large, max amount is 2500 USD
   */
  PAYMENT_AMOUNT_TOO_LARGE_MAX_2500_USD = 'R114',
  /**
   * POS number is invalid
   */
  POS_NUMBER_IS_INVALID = 'R115',
  /**
   * Company is not assigned
   */
  COMPANY_IS_NOT_ASSIGNED = 'R117',
  /**
   * Balance lookup issue
   */
  BALANCE_LOOKUP_ISSUE = 'R121',
  /**
   * Balance not yet available for this biller
   */
  BALANCE_NOT_YET_AVAILABLE_FOR_THIS_BILLER = 'R122',
  /**
   * Processor OK
   */
  PROCESSOR_OK = 'R200',
  /**
   * Processor Access Denied
   */
  PROCESSOR_ACCESS_DENIED = 'R201',
  /**
   * Processor does not return balance
   */
  PROCESSOR_DOES_NOT_RETURN_BALANCE = 'R202',
  /**
   * Catchall Processor error
   */
  CATCHALL_PROCESSOR_ERROR = 'R299',
}
