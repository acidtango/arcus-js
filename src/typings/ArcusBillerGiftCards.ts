import { ArcusAmountList } from './ArcusAmountList';
import { ArcusRange } from './ArcusRange';
import { ArcusBiller } from './ArcusBiller';
import { BillType } from './ArcusBillType';
import { ArcusType } from './ArcusType';

export type ArcusBillerGiftCards = ArcusBiller & {
  /**
   * Every gift card biller type will be biller
   */
  type: ArcusType.BILLER;
  /**
   * Every gift card billerâ€™s bill type will be GiftCard
   * @example "GiftCard"
   */
  billerType: 'GiftCard';
  /**
   * Every gift card biller will be of type account_number.
   */
  billType: BillType;
  /**
   * A value or range of US dollar gift card amounts available for a biller.
   */
  availableGiftCardAmounts: ArcusAmountList | ArcusRange;
  /**
   * Commission applied by the carrier.
   */
  giftCardCommission: number;
};
