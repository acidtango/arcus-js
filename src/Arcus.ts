import fetch, { Response } from 'node-fetch';
import { ArcusAccount } from './typings/ArcusAccount';
import { ArcusBill } from './typings/ArcusBill';
import { ArcusBillCreate } from './typings/ArcusBillCreate';
import { ArcusBillerGiftCards } from './typings/ArcusBillerGiftCards';
import { ArcusBillerTopUp } from './typings/ArcusBillerTopUp';
import { ArcusBillerUtility } from './typings/ArcusBillerUtility';
import { ArcusBillPaymentCreateResponse } from './typings/ArcusBillPaymentCreateResponse';
import { ArcusBillXData } from './typings/ArcusBillXData';
import { ArcusError } from './typings/ArcusError';
import { ArcusErrorCode } from './typings/ArcusErrorCode';
import { ArcusErrorResponse } from './typings/ArcusErrorResponse';
import { ArcusGetTransactionParams } from './typings/ArcusGetTransactionParams';
import { ArcusSingleConsultParams } from './typings/ArcusSingleConsultParams';
import { ArcusSinglePayParams } from './typings/ArcusSinglePayParams';
import { ArcusTransaction } from './typings/ArcusTransaction';
import { CreateBillPayment } from './typings/CreateBillPayment';
import { GetBillersGiftCardsResponse } from './typings/GetBillersGiftCardsResponse';
import { GetBillersResponse } from './typings/GetBillersResponse';
import { GetBillersTopUpsResponse } from './typings/GetBillersTopUpsResponse';
import { GetBillerUtilitiesResponse } from './typings/GetBillerUtilitiesResponse';
import { GetBillsListResponse } from './typings/GetBillsListResponse';
import { GetTransactionsListResponse } from './typings/GetTransactionsListResponse';
import { PostSingleConsultResponse } from './typings/PostSingleConsultResponse';
import { camelize } from './utils/camelize';
import { ArcusCrypto } from './utils/crypto/ArcusCrypto';
import { ArcusCryptoNode } from './utils/crypto/ArcusCryptoNode';
import { generateBody } from './utils/generateBody';

/**
 * Arcus Fintech JS client
 *
 * API Reference: https://docs.arcusfi.com/api/3.mx/endpoints/
 */
export class Arcus {
  public static create(apiKey: string, secretKey: string) {
    return new Arcus(apiKey, secretKey, fetch, new ArcusCryptoNode());
  }

  private static extractJson(response: Response) {
    return response.json();
  }

  private static async dealWithErrors(response: Response) {
    if (response.status >= 400) {
      const body: ArcusErrorResponse | undefined = await response.json();

      throw new ArcusError(
        body?.code || ArcusErrorCode.UNEXPECTED_ERROR,
        body?.message || `Unknown error from Arcus: ${response.status}`,
      );
    }
    return response;
  }

  private config: {
    baseURL: string;
    headers: { Accept: string; 'Content-Type': string; 'Content-MD5': string };
  };

  constructor(
    private apiKey: string,
    private secretKey: string,
    private http: typeof fetch,
    private crypto: ArcusCrypto,
  ) {
    this.config = {
      baseURL: 'https://api.staging.arcusapi.com',
      headers: {
        Accept: 'application/vnd.regalii.v3.mx+json',
        'Content-Type': 'application/json',
        'Content-MD5': '',
      },
    };
  }

  private computeChecksum(path: string, date: string) {
    const text = `application/json,,${path},${date}`;
    return this.crypto.hmacsha1(this.secretKey, text);
  }

  private generateSignatureHeaders(path: string, date: Date = new Date()) {
    const dateStr = date.toUTCString();

    return {
      Date: dateStr,
      Authorization: `APIAuth ${this.apiKey}:${this.computeChecksum(path, dateStr)}`,
    };
  }

  private generateDefaultHeaders(path: string) {
    return {
      ...this.config.headers,
      ...this.generateSignatureHeaders(path),
    };
  }

  getAccount(): Promise<ArcusAccount> {
    const path = '/account';

    return this.http(this.config.baseURL + path, {
      method: 'GET',
      headers: this.generateDefaultHeaders(path),
    })
      .then(Arcus.dealWithErrors)
      .then(Arcus.extractJson)
      .then((data) => camelize<ArcusAccount>(data));
  }

  getBillers() {
    const path = '/billers';

    return this.http(this.config.baseURL + path, {
      method: 'GET',
      headers: this.generateDefaultHeaders(path),
    })
      .then(Arcus.dealWithErrors)
      .then(Arcus.extractJson)
      .then((data) => camelize<GetBillersResponse>(data))
      .then((data) => data.billers);
  }

  getBillersUtilities(): Promise<ArcusBillerUtility[]> {
    const path = '/billers/utilities';

    return this.http(this.config.baseURL + path, {
      method: 'GET',
      headers: this.generateDefaultHeaders(path),
    })
      .then(Arcus.dealWithErrors)
      .then(Arcus.extractJson)
      .then((data) => camelize<GetBillerUtilitiesResponse>(data))
      .then((data) => data.billers);
  }

  getBillersTopUps(): Promise<ArcusBillerTopUp[]> {
    const path = '/billers/topups';

    return this.http(this.config.baseURL + path, {
      method: 'GET',
      headers: this.generateDefaultHeaders(path),
    })
      .then(Arcus.dealWithErrors)
      .then(Arcus.extractJson)
      .then((data) => camelize<GetBillersTopUpsResponse>(data))
      .then((data) => data.billers);
  }

  /**
   * Provides the current list of gift card billers and their associated properties.
   *
   * To ensure that your customers are able to leverage the enhancements that we make over time, you are required to cache the full list of billers and update it on a weekly basis.
   */
  getBillersGiftCards(): Promise<ArcusBillerGiftCards[]> {
    const path = '/billers/gift_cards';

    return this.http(this.config.baseURL + path, {
      method: 'GET',
      headers: this.generateDefaultHeaders(path),
    })
      .then(Arcus.dealWithErrors)
      .then(Arcus.extractJson)
      .then((data) => camelize<GetBillersGiftCardsResponse>(data))
      .then((data) => data.billers);
  }

  getTransactions(params?: ArcusGetTransactionParams): Promise<ArcusTransaction[]> {
    const path = '/transactions';
    const query = params?.externalId ? `/q[external_id_eq]=${params.externalId}` : ``;

    return this.http(this.config.baseURL + path + query, {
      method: 'GET',
      headers: this.generateDefaultHeaders(path),
    })
      .then(Arcus.dealWithErrors)
      .then(Arcus.extractJson)
      .then((data) => camelize<GetTransactionsListResponse>(data))
      .then((data) => data.transactions);
  }

  getBillsList(): Promise<ArcusBill[]> {
    const path = '/bills';

    return this.http(this.config.baseURL + path, {
      method: 'GET',
      headers: this.generateDefaultHeaders(path),
    })
      .then(Arcus.dealWithErrors)
      .then(Arcus.extractJson)
      .then((data) => camelize<GetBillsListResponse>(data))
      .then((data) => data.bills);
  }

  createBill(params: ArcusBillCreate): Promise<ArcusBill> {
    const path = '/bills';

    return this.http(this.config.baseURL + path, {
      body: generateBody(params),
      method: 'POST',
      headers: this.generateDefaultHeaders(path),
    })
      .then(Arcus.dealWithErrors)
      .then(Arcus.extractJson)
      .then((data) => camelize<ArcusBill>(data));
  }

  createBillPayment(billId: string, body: CreateBillPayment) {
    const path = `/bills/${billId}/pay`;

    return this.http(this.config.baseURL + path, {
      body: generateBody(body),
      method: 'POST',
      headers: this.generateDefaultHeaders(path),
    })
      .then(Arcus.dealWithErrors)
      .then(Arcus.extractJson)
      .then((data) => camelize<ArcusBillPaymentCreateResponse>(data));
  }

  postBillRefresh(billId: string) {
    const path = `/bills/${billId}/refresh`;

    return this.http(this.config.baseURL + path, {
      method: 'POST',
      headers: this.generateDefaultHeaders(path),
    })
      .then(Arcus.dealWithErrors)
      .then(Arcus.extractJson)
      .then((data) => camelize<ArcusBill>(data));
  }

  getBillXData(billId: string) {
    const path = `/bills/${billId}/xdata`;

    return this.http(this.config.baseURL + path, {
      method: 'GET',
      headers: this.generateDefaultHeaders(path),
    })
      .then(Arcus.dealWithErrors)
      .then(Arcus.extractJson)
      .then((data) => camelize<ArcusBillXData>(data));
  }

  deleteBill(billId: string) {
    const path = `/bills/${billId}`;

    return this.http(this.config.baseURL + path, {
      method: 'DELETE',
      headers: this.generateDefaultHeaders(path),
    })
      .then(Arcus.dealWithErrors)
      .then(Arcus.extractJson)
      .then((data) => camelize<ArcusBill>(data));
  }

  getBill(billId: string) {
    const path = `/bills/${billId}`;

    return this.http(this.config.baseURL + path, {
      method: 'GET',
      headers: this.generateDefaultHeaders(path),
    })
      .then(Arcus.dealWithErrors)
      .then(Arcus.extractJson)
      .then((data) => camelize<ArcusBill>(data));
  }

  /**
   * This endpoint issues a payment for a bill.
   *
   * _*Dealing with Timeouts*_
   *
   * Timeouts from calling our server are rare but can happen periodically. This is why all calls to `/single/pay` should include an `external_id` parameter that corresponds to a generated unique id in your system, which will be stored with a successfully processed transaction. We recommend use UUID for this field.
   *
   * Also, you can set `pos_number` to identify where this payment is performed. For example: A payment performed in Branch Off ice 12 Cash Register 03 could be described as `S12C03`.
   */
  singlePay(params: ArcusSinglePayParams) {
    const path = '/single/pay';

    return this.http(this.config.baseURL + path, {
      method: 'POST',
      body: generateBody(params),
      headers: this.generateDefaultHeaders(path),
    })
      .then(Arcus.dealWithErrors)
      .then(Arcus.extractJson)
      .then((data) => camelize<ArcusTransaction>(data));
  }

  /**
   * Consult an utility bill in Arcus with an account number. Account number bills are mostly to make payments without providing personal information.
   */
  singleConsult(params: ArcusSingleConsultParams) {
    const path = '/single/consult';

    return this.http(this.config.baseURL + path, {
      method: 'POST',
      body: generateBody(params),
      headers: this.generateDefaultHeaders(path),
    })
      .then(Arcus.dealWithErrors)
      .then(Arcus.extractJson)
      .then((data) => camelize<ArcusTransaction>(data));
  }

  /**
   * This endpoint retrieves a specific transaction.
   */
  getTransaction(transactionId: number) {
    const path = `/transactions/${transactionId}`;

    return this.http(this.config.baseURL + path, {
      method: 'GET',
      headers: this.generateDefaultHeaders(path),
    })
      .then(Arcus.dealWithErrors)
      .then(Arcus.extractJson)
      .then((data) => camelize<ArcusTransaction>(data));
  }

  /**
   * Allows you to cancel a bill payment from the same day for certain billers using the id returned from the bill/pay endpoint. If the bill has already been canceled, the biller does not support canceling, or the allowed time period has already elapsed, the response will contain an error.
   */
  cancelTransaction(transactionId: string) {
    const path = `/transactions/cancel`;

    return this.http(this.config.baseURL + path, {
      method: 'POST',
      body: generateBody({ id: transactionId }),
      headers: this.generateDefaultHeaders(path),
    })
      .then(Arcus.dealWithErrors)
      .then(Arcus.extractJson)
      .then((data) => camelize<PostSingleConsultResponse>(data));
  }
}
