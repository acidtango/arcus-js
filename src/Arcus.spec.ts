import {
  CFE,
  CFE_TRANSACTION,
  NETFLIX,
  TELCEL,
  TOTALPLAY,
  TRANSACTION,
} from '../test/fixtures/fixtures';
import { Arcus } from './Arcus';
import { ArcusBiller } from './typings/ArcusBiller';
import { ArcusErrorCode } from './typings/ArcusErrorCode';

const billerWithName = (name: string) => (utility: ArcusBiller) => utility.name === name;

const billerTotalPlay = billerWithName('TotalPlay');
const billerTelcel = billerWithName('TAETelcel');
const billerNetflix = billerWithName('CÃ“DIGO NETFLIX $150');
const EXAMPLE_FAKE_EXTERNAL_TRANSACTION_ID = 'EXAMPLE_NON_EXISTINT_TRANSACTION';

describe('Arcus', () => {
  const arcus = Arcus.createNew({
    apiKey: process.env.ARCUS_API_KEY as string,
    secretKey: process.env.ARCUS_SECRET_KEY as string,
  });

  it('getAccount', async () => {
    const account = await arcus.getAccount();

    expect(account.balance).toBeDefined();
    expect(account.currency).toEqual(TRANSACTION.currency);
    expect(account.name).toEqual('rabbit');
  });

  it('getBillers', async () => {
    const billers = await arcus.getBillers();

    expect(billers).toHaveLengthWithin(100, 130);

    const totalPlay = billers.find(billerTotalPlay);
    const telcel = billers.find(billerTelcel);
    const netflix = billers.find(billerNetflix);

    expect(totalPlay).toBeDefined();
    expect(totalPlay?.name).toEqual(TOTALPLAY.name);

    expect(telcel).toBeDefined();
    expect(telcel?.name).toEqual(TELCEL.name);

    expect(netflix).toBeDefined();
    expect(netflix?.name).toEqual(NETFLIX.name);
  });

  it('getBillersUtilities', async () => {
    const utilities = await arcus.getBillersUtilities();

    expect(utilities).toHaveLengthWithin(60, 80);

    const totalPlay = utilities.find(billerTotalPlay);

    expect(totalPlay?.name).toEqual(TOTALPLAY.name);
    expect(totalPlay?.billerType).toEqual(TOTALPLAY.billerType);
  });

  it('getBillersTopUps', async () => {
    const topUps = await arcus.getBillersTopUps();

    expect(topUps).toHaveLengthWithin(10, 15);

    const telcel = topUps.find(billerTelcel);

    expect(telcel?.name).toEqual(TELCEL.name);
    expect(telcel?.billerType).toEqual(TELCEL.billerType);
  });

  it('getBillersGiftCards', async () => {
    const giftCards = await arcus.getBillersGiftCards();

    expect(giftCards).toHaveLengthWithin(35, 45);

    const netflix = giftCards.find(billerNetflix);

    expect(netflix?.billerType).toEqual(NETFLIX.billerType);
  });

  describe('getTransactions', () => {
    it('gets Transaction without parameters', async () => {
      const transactions = await arcus.getTransactions();

      expect(transactions).toHaveLengthWithin(95, 105);

      const first = transactions[0];
      expect(first).toBeDefined();
    });

    it('getTransaction With paramater and returns an empty if its not found', async () => {
      const transactions = await arcus.getTransactions({
        externalId: EXAMPLE_FAKE_EXTERNAL_TRANSACTION_ID,
      });
      expect(transactions).toHaveLength(0);
    });

    it('getTransaction With paramater and returns result', async () => {
      const transactions = await arcus.getTransactions({
        externalId: TRANSACTION.externalId,
      });
      expect(transactions).toHaveLength(1);

      const first = transactions[0];

      expect(first.id).toEqual(TRANSACTION.transactionId);
      expect(first.amountCurrency).toEqual(TRANSACTION.currency);
      expect(first.accountNumber).toEqual(TRANSACTION.accountNumber);
    });
  });

  it('getTransaction', async () => {
    const transaction = await arcus.getTransaction(TRANSACTION.transactionId);

    expect(transaction.id).toEqual(TRANSACTION.transactionId);
    expect(transaction.amountCurrency).toEqual(TRANSACTION.currency);
    expect(transaction.accountNumber).toEqual(TRANSACTION.accountNumber);
  });

  it.todo('cancelTransaction');

  it.todo('getBillsList');

  it.todo('postBill');

  it('getBillSingleConsult', async () => {
    const consult = await arcus.singleConsult({
      accountNumber: CFE_TRANSACTION.accountNumber,
      billerId: CFE.billerId,
    });

    expect(consult).toBeDefined();
    expect(consult.accountNumber).toEqual(CFE_TRANSACTION.accountNumber);
    expect(consult.fxRate).toEqual(1);
  });

  it.todo('getBillSinglePay');

  it.todo('getBill');

  it.todo('postBillPayment');

  it.todo('postBillRefresh');

  it.todo('getBillXdata');

  it.todo('deleteBill');

  describe('singlePay', () => {
    // At the time of this test, this biller was failing because a 504, now it's working
    it.skip('does not fail when buying a top up', async () => {
      const error = await arcus
        .singlePay({
          billerId: 13597,
          accountNumber: '4545454545',
          currency: 'MXN',
          amount: 10,
          externalId: 'db8bf3b9-6db5-410f-ac5f-e0967eac5ff9',
        })
        .catch((error) => error);

      expect(error.code).toEqual(ArcusErrorCode.UNEXPECTED_ERROR);
    });
  });
});
