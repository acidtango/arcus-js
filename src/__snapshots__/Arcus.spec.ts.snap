// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Arcus getAccount 1`] = `
Object {
  "balance": 9991243.39,
  "currency": "MXN",
  "minimumBalance": 0,
  "name": "rabbit",
}
`;

exports[`Arcus getBillers 1`] = `
Object {
  "accountNumberDigits": "10",
  "billType": "account_number",
  "billerType": "Cable",
  "canCheckBalance": true,
  "country": "MX",
  "currency": "MXN",
  "hasXdata": false,
  "hoursToFulfill": 0,
  "id": 12493,
  "mask": "9999999999",
  "name": "TotalPlay",
  "requiresNameOnAccount": false,
  "supportsPartialPayments": true,
  "type": "biller",
}
`;

exports[`Arcus getBillers 2`] = `
Object {
  "accountNumberDigits": "10",
  "billType": "phone_number",
  "billerType": "MXCell",
  "canCheckBalance": false,
  "country": "MX",
  "currency": "MXN",
  "hasXdata": false,
  "hoursToFulfill": 0,
  "id": 13603,
  "mask": "(999) 999-9999",
  "name": "TAETelcel",
  "requiresNameOnAccount": false,
  "supportsPartialPayments": false,
  "type": "biller",
}
`;

exports[`Arcus getBillers 3`] = `
Object {
  "accountNumberDigits": "10",
  "billType": "phone_number",
  "billerType": "GiftCard",
  "canCheckBalance": false,
  "country": "MX",
  "currency": "MXN",
  "hasXdata": false,
  "hoursToFulfill": null,
  "id": 13722,
  "mask": "9999999999",
  "name": "CÓDIGO NETFLIX $150",
  "requiresNameOnAccount": false,
  "supportsPartialPayments": false,
  "type": "biller",
}
`;

exports[`Arcus getBillersGiftCards 1`] = `
Object {
  "accountNumberDigits": "10",
  "availableGiftCardAmounts": Array [
    "150.00",
  ],
  "billType": "phone_number",
  "billerType": "GiftCard",
  "canCheckBalance": false,
  "country": "MX",
  "currency": "MXN",
  "giftCardCommission": 3.75,
  "hasXdata": false,
  "hoursToFulfill": null,
  "id": 13722,
  "mask": "9999999999",
  "name": "CÓDIGO NETFLIX $150",
  "requiresNameOnAccount": false,
  "supportsPartialPayments": false,
  "type": "biller",
}
`;

exports[`Arcus getBillersTopUps 1`] = `
Object {
  "accountNumberDigits": "10",
  "availableTopupAmounts": Array [
    "10.00",
    "20.00",
    "30.00",
    "50.00",
    "100.00",
    "150.00",
    "200.00",
    "300.00",
    "500.00",
  ],
  "billType": "phone_number",
  "billerType": "MXCell",
  "canCheckBalance": false,
  "country": "MX",
  "currency": "MXN",
  "hasXdata": false,
  "hoursToFulfill": 0,
  "id": 13603,
  "mask": "(999) 999-9999",
  "name": "TAETelcel",
  "requiresNameOnAccount": false,
  "supportsPartialPayments": false,
  "topupCommission": 7.5,
  "type": "biller",
}
`;

exports[`Arcus getBillersUtilities 1`] = `
Object {
  "accountNumberDigits": "10",
  "billType": "account_number",
  "billerType": "Cable",
  "canCheckBalance": true,
  "country": "MX",
  "currency": "MXN",
  "hasXdata": false,
  "hoursToFulfill": 0,
  "id": 12493,
  "mask": "9999999999",
  "name": "TotalPlay",
  "requiresNameOnAccount": false,
  "supportsPartialPayments": true,
  "type": "biller",
}
`;

exports[`Arcus getTransaction 1`] = `
Object {
  "accountNumber": "256863285",
  "amount": 12.21,
  "amountCurrency": "MXN",
  "amountUsd": 0.61,
  "createdAt": "2021-01-26T09:44:35Z",
  "externalId": "84bce950-c878-483e-aeca-41217f969301",
  "fxRate": 1,
  "hoursToFulfill": 24,
  "id": 100034581,
  "status": "fulfilled",
  "ticketText": "Pago realizado exitosamente Auth: 1611654275",
  "totalUsd": 0.61,
  "transactionFee": 0,
  "type": "transaction",
}
`;

exports[`Arcus getTransactionsList 1`] = `
Object {
  "accountNumber": "256863285",
  "amount": 12.21,
  "amountCurrency": "MXN",
  "amountUsd": 0.61,
  "createdAt": "2021-01-26T09:44:35Z",
  "externalId": "84bce950-c878-483e-aeca-41217f969301",
  "fxRate": 1,
  "hoursToFulfill": 24,
  "id": 100034581,
  "status": "fulfilled",
  "ticketText": "Pago realizado exitosamente Auth: 1611654275",
  "totalUsd": 0.61,
  "transactionFee": 0,
  "type": "transaction",
}
`;
